src/upsample.jl

```diff
@@ -1,4 +1,68 @@
-export upsample_bilinear, ∇upsample_bilinear, pixel_shuffle
+export upsample_nearest, ∇upsample_nearest,
+    upsample_bilinear, ∇upsample_bilinear,
+    pixel_shuffle
+
+"""
+    upsample_nearest(x::AbstractArray, scale::NTuple{S,Int})
+
+Upsamples by integer multiples along the first `S` dimensions.
+Subsequent dimensions of `x` are not altered.
+
+See also [`upsample_bilinear`](@ref), for two dimensions of an `N=4` array.
+
+# Example
+```jldoctest
+julia> upsample_nearest([1 2 3; 4 5 6], (2,3))
+4×9 Array{$Int,2}:
+ 1  1  1  2  2  2  3  3  3
+ 1  1  1  2  2  2  3  3  3
+ 4  4  4  5  5  5  6  6  6
+ 4  4  4  5  5  5  6  6  6
+
+julia> upsample_nearest([1 2 3; 4 5 6], (2,))
+4×3 Array{$Int,1}:
+ 1  2  3
+ 1  2  3
+ 4  5  6
+ 4  5  6
+```
+"""
+function upsample_nearest(x::AbstractArray{T,N}, scales::NTuple{S, <:Integer}) where {T,N,S}
+    S in 1:N || throw(ArgumentError("can't upsample ndims(x)=$N with scale=$scales"))
+    outsize = ntuple(d -> d<=S ? scales[d] * size(x,d) : size(x,d), N)
+    out = similar(x, T, outsize)
+    writesize = ntuple(N+S) do d
+        d > 2S && return size(x, d-S)
+        isodd(d) ? scales[cld(d,2)] : size(x, cld(d,2))
+    end
+    readsize = ntuple(N+S) do d
+        d > 2S && return size(x, d-S)
+        isodd(d) ? 1 : size(x, cld(d,2))
+    end
+    reshape(out, writesize) .= reshape(x, readsize)
+    out
+end
+
+function ∇upsample_nearest(x::AbstractArray{T,N}, scales::NTuple{S, <:Integer}) where {T,N,S}
```

Comment:
x  is supposed to be the input or the incoming sensitivities? If the latter, calling it Δ could be clearer

Reply:
Was going to call it `downsample_nearest` until I realised that would imply `mean` while the gradient wants `sum`. Can change `x` if you think it too confusing.

Reply:
I don't see a precedent for downsample nearest in other libraries, so I think this is fine as it is